samples/web_shell/lib/shell/service_worker.js

```diff
@@ -0,0 +1,168 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import JsFileSystem from 'axiom/fs/js/file_system';
+import DomFileSystem from 'axiom/fs/dom/file_system';
+import GDriveFileSystem from 'axiom/fs/gdrive/file_system';
+import FileSystemManager from 'axiom/fs/base/file_system_manager';
+import ExecuteContext from 'axiom/fs/base/execute_context';
+import StdioSource from 'axiom/fs/stdio_source';
+import Path from 'axiom/fs/path';
+import AxiomError from 'axiom/core/error';
+
+/**
+ * @constructor
+ *
+ * A Service worker class.
+ *
+ * @param FileSystemManager fsm
+ */
+var ServiceWorker = function(fsm) {
+  this.fsm = fsm;
+};
+
+export {ServiceWorker};
+export default ServiceWorker;
+
+/**
+ *
+ * @return Promise<null>
+ */
+ServiceWorker.prototype.register = function() {
+  if ('serviceWorker' in navigator) {
+    return navigator.serviceWorker.register('/service_worker.js',
+        {scope: '/'}).then(function(reg) {
+      // registration worked
+      console.log('Registration succeeded. Scope is ' + reg.scope);
+    }).catch(function(error) {
+      // registration failed
+      console.log('Registration failed with ' + error);
+    });
+  }
+}
+
+/**
+ * @param {string} message
+ *
+ * @return Promise<@>
+ */
+ServiceWorker.prototype.processMessage = function(message) {
+
+  var url = this.getFileSystemUrl(message);
```

Comment:
I would rename this to "path", since we are really returning a Path to use with fsm.
`var path = this.getFileSystemPath(message);`